src/ILCompiler/src/RdXmlRootProvider.cs

```diff
@@ -0,0 +1,128 @@
+            {
+                switch (element.Name.LocalName)
+                {
+                    case "Assembly":
+                        ProcessAssemblyDirective(rootProvider, element);
+                        break;
+
+                    default:
+                        throw new NotSupportedException();
+                }
+            }
+        }
+
+        private void ProcessAssemblyDirective(IRootingServiceProvider rootProvider, XElement assemblyElement)
+        {
+            var assemblyNameAttribute = assemblyElement.Attribute("Name");
+            if (assemblyNameAttribute == null)
+                throw new Exception();
+
+            ModuleDesc assembly = _context.ResolveAssembly(new AssemblyName(assemblyNameAttribute.Value));
+
+            var dynamicDegreeAttribute = assemblyElement.Attribute("Dynamic");
+            if (dynamicDegreeAttribute != null)
+            {
+                if (dynamicDegreeAttribute.Value != "Required All")
+                    throw new NotSupportedException();
+
+                // Reuse LibraryRootProvider to root everything
+                new LibraryRootProvider((EcmaModule)assembly).AddCompilationRoots(rootProvider);
+            }
+
+            foreach (var element in assemblyElement.Elements())
+            {
+                switch (element.Name.LocalName)
+                {
+                    case "Type":
+                        ProcessTypeDirective(rootProvider, assembly, element);
+                        break;
+                    default:
+                        throw new NotSupportedException();
+                }
+            }
+        }
+
+        private void ProcessTypeDirective(IRootingServiceProvider rootProvider, ModuleDesc containingModule, XElement typeElement)
+        {
+            var typeNameAttribute = typeElement.Attribute("Name");
+            if (typeNameAttribute == null)
+                throw new Exception();
+
+            var dynamicDegreeAttribute = typeElement.Attribute("Dynamic");
+            if (dynamicDegreeAttribute != null)
+            {
+                if (dynamicDegreeAttribute.Value != "Required All")
+                    throw new NotSupportedException();
+            }
+
+            string typeName = typeNameAttribute.Value;
+
+            string name;
+            string ns;
+            int namespaceEndIndex = typeName.LastIndexOf('.');
+            if (namespaceEndIndex > 0)
+            {
+                ns = typeName.Substring(0, namespaceEndIndex);
+                name = typeName.Substring(namespaceEndIndex + 1);
+            }
+            else
+            {
+                ns = string.Empty;
+                name = typeName;
+            }
+
+            MetadataType type = containingModule.GetType(ns, name);
+            rootProvider.AddCompilationRoot(type, "RD.XML root");
+
+            foreach (var method in type.GetMethods())
+            {
+                if (method.IsAbstract || method.HasInstantiation)
```

Comment:
Yes, I sidestepped that because I'm not entirely convinced `IsIntrinsic` is what we should be checking (e.g. we have a bunch of methods that are marked intrinsic, but they're not intrinsics for all of our codegen backends). It will need tweaks in the future. I'm hoping that by the time someone hits a problem with this, the RD.XML handling code will be deleted.